### PR TITLE
[Merged by Bors] - chore(Tactic/Ring/Basic): reduce use of autoImplicit

### PR DESCRIPTION
**Pull request by grunweg**
---

- [x] depends on: #14725
- [x] depends on: #14726
- [x] depends on: #14770
- [x] depends on: #14805

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Ring/Basic.lean [hunk 1]
```diff
@@ -282,10 +288,12 @@ inductive Overlap (e : Q($α)) where
   (with nonzero leading coefficient). -/
   | nonzero (_ : Result (ExProd sα) e)
 
-theorem add_overlap_pf (x : R) (e) (pq_pf : a + b = c) :
+-- TODO future! variable {a a₁ a₂ b b₁ b₂ c : R}
```
Comment (jcommelin) [inline Mathlib/Tactic/Ring/Basic.lean:291]: What goes wrong if you do this now?

Comment (grunweg) [reply Mathlib/Tactic/Ring/Basic.lean]: Thanks for pointing out this stray code. It works now (it didn't when `a` and `b` where shadowing some `Q(...)` expressions; these are now all explicit). I've pushed a commit reducing the diff, using this.

File: Mathlib/Tactic/Ring/Basic.lean [hunk 2]
```diff
@@ -957,29 +994,32 @@ def ExSum.evalInv (czα : Option Q(CharZero $α)) (va : ExSum sα a) :
...
+def evalDiv {a b : Q($α)} (rα : Q(DivisionRing $α)) (czα : Option Q(CharZero $α)) (va : ExSum sα a)
     (vb : ExSum sα b) : AtomM (Result (ExSum sα) q($a / $b)) := do
   let ⟨_c, vc, pc⟩ ← vb.evalInv sα rα czα
   let ⟨d, vd, (pd : Q($a * $_c = $d))⟩ := evalMul sα va vc
   pure ⟨d, vd, (q(div_pf $pc $pd) : Expr)⟩
 
-theorem add_congr (_ : a = a') (_ : b = b')
-    (_ : a' + b' = c) : (a + b : R) = c := by subst_vars; rfl
+theorem add_congr
+    (_ : a = a') (_ : b = b') (_ : a' + b' = c) : (a + b : R) = c := by
```
Comment (jcommelin) [inline Mathlib/Tactic/Ring/Basic.lean:1011]:
```suggestion
theorem add_congr (_ : a = a') (_ : b = b') (_ : a' + b' = c) : (a + b : R) = c := by
```

File: Mathlib/Tactic/Ring/Basic.lean [hunk 3]
```diff
@@ -957,29 +994,32 @@ def ExSum.evalInv (czα : Option Q(CharZero $α)) (va : ExSum sα a) :
...
+  subst_vars; rfl
 
-theorem mul_congr (_ : a = a') (_ : b = b')
-    (_ : a' * b' = c) : (a * b : R) = c := by subst_vars; rfl
+theorem mul_congr
+    (_ : a = a') (_ : b = b') (_ : a' * b' = c) : (a * b : R) = c := by
```
Comment (jcommelin) [inline Mathlib/Tactic/Ring/Basic.lean:1015]:
```suggestion
theorem mul_congr (_ : a = a') (_ : b = b') (_ : a' * b' = c) : (a * b : R) = c := by
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by grunweg** [comment]
Thanks for the fast review!
Let's try: does `bors merge` also work for me?
bors merge

**Pull request closed without merging**